### PR TITLE
xt-clang: support bundled newlib

### PR DESCRIPTION
**Pull request by dcpleung**
This PR contains commits to support the (very old) newlib bundled with Xtensa LLVM toolchain (xt-clang). See individual commits for details.

### CODE REVIEW

**Comment by dcpleung** [comment]
Compliance check failure is due to multi-line #if where the second line starts with spaces.

**Comment by dcpleung** [comment]
The newlib in xt-clang is too broken to support Zephyr. So closing it for now...

File: include/zephyr/posix/sys/stat.h [hunk 1]
```diff
@@ -34,10 +34,29 @@
...
 #endif
 
+#if defined(__XT_CLANG__) && defined(CONFIG_NEWLIB_LIBC)
+/*
+ * Xtensa xt-clang is based on old newlib which does not _timespec.h.
+ * And some of the typedefs below have already been defined, but
+ * there are no corresponding *_DECLARED macros, so we need to define
+ * them manually to avoid them being re-declared.
+ */
+#define HAS__TIMESPEC_H 0
```
Comment (keith-packard) [inline include/zephyr/posix/sys/stat.h:44]: Vague concerns about a name in the 'application' namespace, but prefixing it with '_' would risk conflicting with a C library define.

File: lib/libc/newlib/libc-hooks.c [hunk 2]
```diff
@@ -23,6 +23,8 @@
 #include <zephyr/sys/mem_manage.h>
 #include <sys/time.h>
 
+#include <newlib.h>
```
Comment (keith-packard) [inline lib/libc/newlib/libc-hooks.c:26]: Surely one of the above headers already includes this? Odd, if not. But, nbd in any case.

File: lib/libc/newlib/libc-hooks.c [hunk 3]
```diff
@@ -303,10 +305,8 @@ void *_sbrk(intptr_t count)
 }
 __weak FUNC_ALIAS(_sbrk, sbrk, void *);
 
-#ifdef CONFIG_MULTITHREADING
```
Comment (keith-packard) [inline lib/libc/newlib/libc-hooks.c:306]: Uh, this isn't good -- what does this old version of newlib do for locking? Seems like this will cause all kinds of trouble if multiple threads call into stdio or anything else in newlib that requires locking.

File: lib/libc/newlib/libc-hooks.c [hunk 4]
```diff
@@ -450,7 +450,21 @@ void __retarget_lock_release_recursive(_LOCK_T lock)
 	__ASSERT_NO_MSG(lock != NULL);
 	k_mutex_unlock((struct k_mutex *)lock);
 }
-#endif /* CONFIG_MULTITHREADING */
+#else /* CONFIG_MULTITHREADING && _RETARGETABLE_LOCKING */
```
Comment (keith-packard) [inline lib/libc/newlib/libc-hooks.c:453]: Do you want to look into sharing the common malloc implementation instead of using the newlib one? I realize it's not directly aligned with your goals here, but I don't know anyone else poking at newlib code...

File: subsys/logging/log_output_syst.c [hunk 5]
```diff
@@ -751,11 +751,11 @@ static int mipi_catalog_formatter(cbprintf_cb out, void *ctx,
 		}
 
 		if (arg_sz == sizeof(mipi_syst_u64)) {
-			*((mipi_syst_u64 *)argp) =
-				(mipi_syst_u64)MIPI_SYST_HTOLE64(val.v64);
+			val.v64 = MIPI_SYST_HTOLE64(val.v64);
```
Comment (keith-packard) [inline subsys/logging/log_output_syst.c:754]: This looks like a bug fix for unaligned access here -- seems like it should be in a separate PR?

**Review by keith-packard — COMMENTED** [review]
I've got a couple of questions, but no major blockers here. Thanks for figuring this out.

**Comment by dcpleung** [comment]
From what I can tell, the newlib in xt-clang is very old and that there are various issues... I think we should not re-introduce the old behavior here. Given that we have picolibc, we should be using that instead, IMHO.

**Comment by keith-packard** [comment]
Yeah, iirc, you had some adventures with that too. Good luck!

**Comment by dcpleung** [comment]
Need to re-open this... this is needed to use the profiling tools provided by Cadence.

File: lib/libc/newlib/libc-hooks.c (hunk 3 above)
Comment (dcpleung) [reply lib/libc/newlib/libc-hooks.c:306]: From what I have read, old newlib is not really thread safe. That's why they introduced the retargetable locks.

File: lib/libc/newlib/libc-hooks.c (hunk 4 above)
Comment (dcpleung) [reply lib/libc/newlib/libc-hooks.c:453]: I cannot find a definite way to exclude a function from a static library during linking. IIRC, providing multiple implementation may result in the linker either complaining, or picking one depending on linking order. If it is latter, cmake scripts may need to be changed to ensure ordering all the time.
